components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -73,24 +90,139 @@ class GitpodTerminalService(private val project: Project) : Disposable {
 
         messageBusConnection.subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
 
-        return completableFuture
+        completableFuture.get()
+    }
+
+    private tailrec fun waitUntilGuestClientIsConnected() {
+        val project = RestService.getLastFocusedOrOpenedProject()
+        var session: ClientSession? = null
+        if (project != null) {
+            session = ClientSessionsManager.getProjectSessions(project, false).firstOrNull()
+        }
+        if (session == null) {
+            session = ClientSessionsManager.getAppSessions(false).firstOrNull()
+        }
+        if (session != null) {
+            return
+        } else {
+            TimeUnit.SECONDS.sleep(1)
+            return waitUntilGuestClientIsConnected()
+        }
     }
 
-    private suspend fun delayUntilControllerClientConnects() {
-        while (UnattendedHostManager.getInstance().controllerClientId == null) {
-            delay(1000L)
+    private fun createSharedTerminalAndExecuteCommand(name: String, command: String) {
+        val registeredTerminals = terminalView.widgets.toMutableList()
+
+        backendTerminalManager.createNewSharedTerminal(UUID.randomUUID().toString(), name)
```

Comment:
JetBrains team suggested we check `JBTerminalWidget.TerminalTitle.change` for changing the terminal tab title. I'm gonna try it now.

Reply:
I've updated the code.
It worked but will only update the title after the terminal window gains focus.